### PR TITLE
The great refactor pt1

### LINKED ISSUE #90: Reorganization of folders and files
- [x] Remove double-nesting of folders (e.g., `discord_quote/discord_quote`)
- [x] Separate out each command into it's own file?
     - ~~[ ] !get, !put, !list and !delete should all be placed in one file~~
     - ~~[ ] !test should be placed in one file~~
     - ~~[ ] !quote, _bot_quote, _hook_quote should be placed in one file~~
     - [x] Any request parsing/cleaning functions that will be shared between commands should be placed in a `utils.py`

_Note:_ This was originally part of #76

### PR DESCRIPTION
**Pull request by synsypa**
# Summary

This PR is mainly to reorganize the package so that we can incrementally refactor the quotebot in future PRs.

# Overview
This aspect of the refactor contains:

- Creates a more sane directory structure for everything
    - `src` contains all the code: `src/bot` is bot related items, `src/author_model` is author model related
    - `bin` contains all the data: most notably the AuthorNet checkpoint
- `discord_quote.py` should now **only* *include bot commands. This should both make the main bot script cleaner (it should primarily import relevant functions and tools as well as make it clearer what functions interact which what different quote calls
    - message processing or otherwise uncategorized utiltiy functions exist in `src/bot/utils.py`
    - database interaction utilities exist in `src/bot/db.py`
    - webhook quote functions exist in `src/bot/webhook_quote.py`
    - traditional quote functions exist in `src/bot/quote.py`
- `utils.py` contains generic helper functions for parsing requests and for cleaning up requests, since we do this often across commands
- Additionally:
    - Moves all legacy functions to `legacy_discord_quote.py`
    - Creates a minimal `discord_quote.py` (mostly functional)
    - Replace `logging` with `logzero`

## To-Do
- [x] ~~No command in `discord_quote.py` should manually implement the logic to choose between webhook vs traditional quoting. there should be a separate abstraction (e.g. `def quote()`)of all the logic in i.e. https://github.com/play4honor/discord_quote_bot/blob/8c70f6530ad71bef96ae38cae85fabe64e180dc7/src/bot/discord_quote.py#L181-L199~~
    - The design of this will be tackled in a separate PR and has been moved to issue #85 
- [x] ~~Similarly, it may be better design to abstract as much logic into separate functions as possible for ease of iteration. e.g. rather than having `!misquote` construct the entire message in that bot function, have it call `construct_direct_misquote` and `construct_predicted_misquote` and just return a single string that can be passed to `quote`~~
    - ~~The paradigm i'm thinking here is basically a functional programming one~~
    - The design of this will be tackled in a separate PR and has been moved to issue #85 
- [x] The Dockerfile needs to be changed to properly reflect the new file structure (maybe)
- [x] Many of the utility functions send things to the logger, but none of them actually receive the logger as an argument, i'm not sure what the correct design pattern here is. My instinct is to let pylint be upset, since I don't think it makes sense to have to pass the logger back and forth
    - (See https://github.com/play4honor/discord_quote_bot/pull/84#issuecomment-697260340)
- [x] NONE OF THIS IS TESTED
    - All functionality except for `!misquote` is minimally working
- [x] ~~### REALLY THOUGH ITS NOT TESTED AT ALL~~
- [x] Setup your own personal testing bot (see the README)
    - You need to be able to test your changes. I do not want to be your tester.
- [x] Restore misquote functionality
    - I've commented out both the import and the legacy function for now, because it was not importing in my venv.
- [x] Fix the `parse_request()` function
    - I jerry-rigged this so that it wouldn't give me errors but there is some list/string confusion (which is causing the spacing issues.
    - image
- [x] Am I importing the packages in a correct way? https://github.com/play4honor/discord_quote_bot/pull/84/files#r493388979

## Pre-Merge Checklist

- [x] Does it work in a personal testing bot?
- [x] Has the docker-file paths been edited?

### CODE REVIEW

**Comment by zhangchuck** [comment]
- I'd call the `quote()` function something a bit different. Maybe "speak()" or some variant of "send()". Whenever the bot is *sending* a message, it should call this function rather than any of the myriad ways we can send via the bot.
     **Suggested function signature**:
     **f(`guild_id`, `channel_id`, `message`, `embeds` (optional)) --> Boolean** (success indicator)
- What should be the function signature of the abstract `parse_command()` function?
     **Suggested function signature**: 
     **f(`ctx`, `invoked_command`, `string`) --> Dict(Guild, Channel, Author, OriginMsg,TargetMsg, Reply, InvokedCommand)**
    `invoked_command` helps the `parse_command()` function decide how to parse the command

**Comment by zhangchuck** [comment]
On second thought, maybe the first step is to *just* implement the de-tangling of files. e.g., just to fix folder-structure

**Comment by synsypa** [comment]
the current message parsing is implemented as a barely edited copy/paste from what used to exist:
https://github.com/play4honor/discord_quote_bot/pull/84/files#diff-0d8dd59df080bc85a83a78768d0f4ebeR54-R78

**Comment by zhangchuck** [comment]
We should tag the current version as release v0.1, or something.

**Comment by zhangchuck** [comment]
Also, maybe we should move the internals of all of our commands into a `legacy.py` file (and call on the `legacy.quote()` etc.) so we can rewrite each command onto the new architecture separately instead of trying to rewrite all at once..

**Comment by zhangchuck** [comment]
## Additional Edits (CZ)

- Switched from default `logging` to `logzero` in most files
- `discord_quote.py` has been stripped bare so that it only contains direct commands
    - to maintain functionality (until we replace in the refactor), we have created `legacy_discord_quote.py`, which contains all of the v0.1 versions of the commands
    - `discord_quote.py` imports these functions and (currently) directly passes along the needed arguments
    - functionality is similar but not perfect

File: src/bot/legacy_discord_quote.py [hunk 1]
```diff
@@ -4,153 +4,27 @@
...
-fmt = logging.Formatter(u'\u241e'.join(['%(asctime)s',
-                                        '%(name)s',
-                                        '%(levelname)s',
-                                        '%(funcName)s',
-                                        '%(message)s']))
-streamInstance = logging.StreamHandler(stream=sys.stdout)
-streamInstance.setFormatter(fmt)
-log.addHandler(streamInstance)
-log.setLevel(logging.DEBUG)
+#import src.author_model.author_model as author
```
Comment (zhangchuck) [inline src/bot/legacy_discord_quote.py]: This was failing!

File: src/bot/legacy_discord_quote.py [hunk 2]
```diff
@@ -294,242 +149,7 @@ async def quote(ctx, *, request:str):
...
-
-    log.info(log_msg(['formatted_quote', ' '.join(reply)]))
-
-    await ctx.channel.send(output)
-
-    log.info(log_msg(['sent_message', 'quote', ctx.message.channel.name]))
-
-@bot.command()
-async def misquote(ctx , *target : discord.User):
+""" async def misquote(ctx , *target : discord.User):
```
Comment (zhangchuck) [inline src/bot/legacy_discord_quote.py]: Since the import was failing, I commented out this entire function.

File: src/bot/discord_quote.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                'channel_join',
+                '\\'.join([channel.guild.name, channel.name])]
+                )
+            )
+
+#            await channel.send('yo we in there')
+
+@bot.command()
+async def me(ctx, *text : str):
+    """
```
Comment (zhangchuck) [inline src/bot/discord_quote.py:70]: Needs a docstring.

File: src/bot/discord_quote.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+async def quote(ctx, *, request:str):
+    """
+    Quotes an existing message from the same channel.
+    request = (MessageID|MessageURL)
+    (Make sure you have Discord's developer mode turned on to get Message IDs)
+    """
+    await old.quote(ctx, request=request, bot=bot)
+
+@bot.command()
+async def misquote(ctx , *target : discord.User):
```
Comment (zhangchuck) [inline src/bot/discord_quote.py:84]: Needs a docstring, also we need to link this over to `old.misquote()`

**Comment by zhangchuck** [comment]
# To-Do's (@synsypa and @DoubleSec)

- [x] Setup your own personal testing bot (see the README)
    - You need to be able to test your changes. I do not want to be your tester.
- [x] Restore misquote functionality
    - I've commented out both the import and the legacy function for now, because it was not importing in my venv.
- [ ] Fix the `parse_request()` function
    - I jerry-rigged this so that it wouldn't give me errors but there is some list/string confusion (which is causing the spacing issues.
    - image
- [ ] Edit the docker file need to be edited for this to successfully deploy to `quote-bot-dev`
- [x] Am I importing the packages in a correct way?

File: src/bot/discord_quote.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+# AWS Packages
+import boto3
+import botocore
+
+# Other Packages
+from logzero import logger as log
+import sys
+import os
+
+sys.path.append('../../')
```
Comment (zhangchuck) [inline src/bot/discord_quote.py]: I had to do this to jerry-rig it properly. Not sure if it's a venv problem or something else.

Recall that we deploy by doing `python3 discord_quote.py` (our working directory is `src/bot/`.

**Comment by zhangchuck** [comment]
> Many of the utility functions send things to the logger, but none of them actually receive the logger as an argument, i'm not sure what the correct design pattern here is. My instinct is to let pylint be upset, since I don't think it makes sense to have to pass the logger back and forth

There's some weird waterfalling that goes on with logger and not sure exactly how it works. For now, I've added:

```
from logzero import logger as log
```

which allows everything to run fine.

File: src/bot/discord_quote.py (hunk 5 above)
Comment (synsypa) [reply src/bot/discord_quote.py]: yea, i haven't done the change but i think my plan is to change the dockerfile to use the base directory as its wd, and then deploy with `python3 src/bot/discord_quote.py`. I can make the adjustments.

**Comment by synsypa** [comment]
i think i fixed the loading thing, it was an issue with a bad file path + some annoying `pickle` behavior from `torch.load()`.

Also fixed up the dockerfile and made some make targets. Didn't have a chance to test yet.

File: Dockerfile [hunk 6]
```diff
@@ -8,10 +8,14 @@ RUN apt-get -y update && \
     apt-get -y autoremove
 
 # Copy the latest version of the bot 
-COPY ./discord_quote /discord_quote
+COPY ./src /src
+COPY ./bin /bin
```
Comment (zhangchuck) [inline Dockerfile]: You're probably just mirroring what we did before, but any idea why we need to do this step?

**Comment by synsypa** [comment]
Tested locally off a docker container, everything should be working.

File: src/bot/discord_quote.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
+# Discord Packages
+import discord
+from discord.ext import commands
+
+# AWS Packages
+import boto3
+import botocore
+
+# Other Packages
+from logzero import logger as log
```
Comment (zhangchuck) [inline src/bot/discord_quote.py:10]: Are we generally okay with this? Or, should we just use

```
from logzero import logger
```

File: src/bot/webhook_quote.py [hunk 8]
```diff
@@ -2,6 +2,8 @@
 import arrow
 import re
 
+from typing import Union
```
Comment (zhangchuck) [inline src/bot/webhook_quote.py:5]: What's this do?

**Comment by zhangchuck** [comment]
London Gents Tussle Messily

**Pull request merged**